### PR TITLE
ci: azure-pipelines

### PR DESCRIPTION
**Pull request by evilebottnawi**
No issue template - azure pipelines

### CODE REVIEW

**Comment by evilebottnawi** [comment]
/cc @hudochenkov can you give me right on settings in repo? Need solve:
`The Azure Pipelines GitHub App is not installed to repository stylelint/stylelint. Learn more`

**Comment by hudochenkov** [comment]
@evilebottnawi done!

We might not have this strange AppVeyor issues, when tests sometimes fails, where it shouldn't.

**Comment by evilebottnawi** [comment]
Before merge:
1. Need investigate why we use `https://webhooks.gitter.im/e/d763493612da45967361` for travis (it is 404 right now)
2. Add `Required` for all azure CIs
3. Add core developments in https://dev.azure.com/stylelint/stylelint organization (need email)

/cc @stylelint/core 

For restarting CI you can use close/open buttons

File: lib/__tests__/ignore.test.js [hunk 1]
```diff
@@ -327,12 +327,7 @@ describe("using ignoreFiles with input files that would cause a postcss syntax e
   });
 
   it("syntax-error-ignored.scss found", () => {
-    // For node 8 on Windows
-    const index = Number(
-      results[1].source.indexOf("syntax-error-ignored.scss")
-    );
-
-    expect(index).not.toBe(-1);
+    expect(results[1].source.indexOf("syntax-error-ignored.scss")).not.toBe(-1);
```
Comment (evilebottnawi) [inline lib/__tests__/ignore.test.js]: /cc @hudochenkov due this lines we have failed CI on windows at random time, i don't investigate why it is happens, but this change fix problem

**Comment by hudochenkov** [comment]
Looks interesting. I like that it's uses Checks tab on Github.

> Need investigate why we use https://webhooks.gitter.im/e/d763493612da45967361 for travis (it is 404 right now)

Screenshot 2019-06-01 at 21 07 29

> Add `Required` for all azure CIs

What does it mean?

> Add core developments in https://dev.azure.com/stylelint/stylelint organization (need email)

Let's use private discussion on Github for this. I've started one.

> For restarting CI you can use close/open buttons

I would avoid that. First, it's a hack and misuse of functionality. Second, it would notify a lot of people by email with no valid reason.

---

Can we change to **Read** access to code? CI doesn't need **write** access.

Screenshot 2019-06-01 at 21 16 34

---

Why there are “errors” on Windows, while there are no errors and tests passed?

Screenshot 2019-06-01 at 20 48 08

Screenshot 2019-06-01 at 20 47 37

File: lib/__tests__/ignore.test.js (hunk 1 above)
Comment (hudochenkov) [reply lib/__tests__/ignore.test.js]: We did have exactly this and had random fails as well https://github.com/stylelint/stylelint/pull/3810. I think we can use:

```js
expect(results[1].source.includes("syntax-error-ignored.scss")).toBe(true);
```

**Comment by ntwb** [comment]
> > For restarting CI you can use close/open buttons

> I would avoid that. First, it's a hack and misuse of functionality. Second, it would notify a lot of people by email with no valid reason.

Agreed, to perform this correctly does each member of the team require a Microsoft account?

> Need investigate why we use https://webhooks.gitter.im/e/d763493612da45967361 for Travis (it is 404 right now)

I think this from the _legacy_ Travis CI integration and that we piped results into Gitter, most likely this can be removed

> Add core developments in dev.azure.com/stylelint/stylelint organization (need email)

We have a dedicated stylelint email address we can use, I'll send this to you privately @evilebottnawi (I have sent it via DM using Gitter.im)

Edit: I have added my email address to the private @stylelint/core discussion, if anyone else would like access please add your email address to the discussion.

-----

Looking at https://dev.azure.com/stylelint/stylelint/_build/results?buildId=7

Each of the Windows build jobs show that they pass, though each of the build jobs has a display of the following, is this a build error, a test error, or an Azure Pipelines bug?
Screen Shot 2019-06-01 at 10 34 02 pm

```shell
❌ PASS lib/rules/value-keyword-case/__tests__/index.js
❌ PASS lib/rules/selector-max-empty-lines/__tests__/index.js
❌ PASS lib/rules/selector-attribute-operator-space-before/__tests__/index.js
❌ PASS lib/rules/selector-attribute-operator-space-after/__tests__/index.js
❌ PASS lib/rules/selector-no-qualifying-type/__tests__/index.js
❌ PASS lib/rules/at-rule-empty-line-before/__tests__/index.js
❌ PASS lib/rules/selector-attribute-brackets-space-inside/__tests__/index.js
❌ PASS lib/rules/selector-pseudo-class-case/__tests__/index.js
❌ PASS lib/utils/parseCalcExpression/__tests__/parseCalcExpression.test.js
❌ PASS lib/rules/property-case/__tests__/index.js
```

Even though each test states it as a `pass` the ❌ does not inspire confidence, in fact I'm not sure what this actually means, did the tests fail? if so why do they state `pass`, if the tests did not fail why are they displayed in red with the ❌ mark?

**Comment by ntwb** [comment]
Another follow up question @evilebottnawi:

Do we need to test on macOS? For the most part it is just a linux box in of itself

File: package.json [hunk 2]
```diff
@@ -100,6 +100,7 @@
     "flow-typed": "^2.5.1",
     "husky": "^2.0.0",
     "jest": "^24.7.1",
+    "jest-junit": "^6.4.0",
```
Comment (ntwb) [inline package.json]: Why is this part of this PR?

Comment (evilebottnawi) [reply package.json]: To integrate azure test output with github, you can see what tests failed or not

**Comment by evilebottnawi** [comment]
> Do we need to test on macOS? For the most part it is just a linux box in of itself

Based on my experience - yes. Some modules like glob/minimatch/etc working differently on macOS, also it is fast and we can honestly say that everything works as expected

**Comment by evilebottnawi** [comment]
@hudochenkov @ntwb 
> What does it mean?

`continuous-integration/travis-ci Expected — Waiting for status to be reported
Required` (need disable Required on travis and appveyor, and add this for azure CI )

> Let's use private discussion on Github for this. I've started one.

:+1: 

> I would avoid that. First, it's a hack and misuse of functionality. Second, it would notify a lot of people by email with no valid reason.

Unfortunately, we can use only 5 users who can restart CI (free version of azure pipelines), maybe in future azure will have better integration with github, but maybe i am wrong, need open issue and get feedback on this

> Can we change to Read access to code? CI doesn't need write access.

Yes we can change this on READ, azure get all permissions for automatically CI releases, it is really awesome feature, but i think we don't need this here and right now

> Each of the Windows build jobs show that they pass, though each of the build jobs has a display of the following, is this a build error, a test error, or an Azure Pipelines bug?

Yes it is bug with jest and azure pipelines, i think it will be fixed in near future (maybe we should open issue). Anyway tests work as expected - if jest exit with -1 test failed, if not all ok, it is just visual bug

File: package.json (hunk 2 above)
Comment (ntwb) [reply package.json]: Thanks 👍🏼

**Comment by ntwb** [comment]
> Yes it is bug with jest and azure pipelines, i think it will be fixed in near future (maybe we should open issue). Anyway tests work as expected - if jest exit with -1 test failed, if not all ok, it is just visual bug

at the moment they appear to be partially working better than when I asked the question above, the errors displayed are still the same errors I mentioned above, but at least the CI build now shows as failed, the above image I posted showed the failures but the CI reported the jobs as passed.

Previously:
Screen Shot 2019-06-01 at 10 34 02 pm

Current:
Screen Shot 2019-06-03 at 4 28 29 pm

Yes each if the test failures continue to show `PASS` on the test, for example: `❌ PASS lib/rules/property-case/__tests__/index.js`

**Comment by evilebottnawi** [comment]
@ntwb I can't understand problem, it is just visual bug between jest and azure pipelines, i can investigate this, you can click on output and see detailed output and you can see what we have problem on windows

**Comment by evilebottnawi** [comment]
Windows failed because we have some bug in stylelint and should fix it, i will investigate. 
Screenshot about red color is just visual bug, jest still works as expected.

**Comment by evilebottnawi** [comment]
Bug here:
```
   Snapshot name: `outputFile writes the result file  1`

    - Snapshot
    + Received

      "
      stylesheet.css
    -  1:3  ✖  Unexpected empty block   block-no-empty
    +  1:3  ×  Unexpected empty block   block-no-empty
      
      "

      48 |       return pify(fs.readFile)(directionPath, "utf-8").then(content => {
      49 |         expect(content).toEqual(systemTestUtils.stripColors(stdout));
    > 50 |         expect(content).toMatchSnapshot();
         |                         ^
      51 |         don
##[error]e();
      52 |       });
      53 |     });

      at toMatchSnapshot (system-tests/outputFile/outputFile.test.js:50:25)


Snapshot Summary
 › 1 snapshot failed from 1 test suite. Inspect your code changes or run `npm run jest -- -u` to update them.
```

**Comment by evilebottnawi** [comment]
eslint migrate on azure pipelines too :wink:

**Comment by jeddy3** [comment]
Closing in favour of https://github.com/stylelint/stylelint/pull/4285

**Pull request closed without merging**